### PR TITLE
[BEAM-14471] Adding testcases and examples for xlang Python DataframeTransform

### PR DESCRIPTION
**Pull request by ihji**
* adding `DataframeTransformTest`
* adding `PythonDataframeWordCount` example
* removing `UsesCrossLanguageTransform` annotation, adding `UsesPythonExpansionService`, `UsesJavaExpansionService` annotations

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [ ] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] Update `CHANGES.md` with noteworthy changes.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/#make-reviewers-job-easier).

To check the build health, please visit [https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md](https://github.com/apache/beam/blob/master/.test-infra/BUILD_STATUS.md)

GitHub Actions Tests Status (on master branch)
------------------------------------------------------------------------------------------------
[Build python source distribution and wheels](https://github.com/apache/beam/actions?query=workflow%3A%22Build+python+source+distribution+and+wheels%22+branch%3Amaster+event%3Aschedule)
[Python tests](https://github.com/apache/beam/actions?query=workflow%3A%22Python+Tests%22+branch%3Amaster+event%3Aschedule)
[Java tests](https://github.com/apache/beam/actions?query=workflow%3A%22Java+Tests%22+branch%3Amaster+event%3Aschedule)

See [CI.md](https://github.com/apache/beam/blob/master/CI.md) for more information about GitHub Actions CI.

### CODE REVIEW

**Comment by ihji** [comment]
Run xvr_flink postcommit

**Comment by ihji** [comment]
Run xvr_spark postcommit

**Comment by ihji** [comment]
run xvr_flink postcommit

**Comment by ihji** [comment]
run xvr_spark postcommit

**Comment by ihji** [comment]
run xvr_javausingpython_dataflow postcommit

**Comment by ihji** [comment]
run xvr_javausingpython_dataflow postcommit

**Comment by ihji** [comment]
run xvr_flink postcommit

**Comment by ihji** [comment]
Run Java PreCommit

**Comment by ihji** [comment]
Run Python PreCommit

**Comment by ihji** [comment]
run xvr_flink postcommit

**Comment by ihji** [comment]
Run xvr_spark postcommit

**Comment by ihji** [comment]
Run xvr_javausingpython_dataflow postcommit

**Comment by ihji** [comment]
Run java precommit

**Comment by ihji** [comment]
Run xvr_spark postcommit

**Comment by ihji** [comment]
R: @chamikaramj

**Comment by ihji** [comment]
run xvr_spark postcommit

**Comment by ihji** [comment]
run xvr_flink postcommit

**Comment by ihji** [comment]
Run SQL PreCommit

**Comment by ihji** [comment]
run xvr_spark postcommit

**Comment by ihji** [comment]
run xvr_flink postcommit

**Comment by ihji** [comment]
run xvr_spark postcommit

**Comment by ihji** [comment]
Run Java PreCommit

**Comment by chamikaramj** [comment]
Run XVR_Direct PostCommit

File: buildSrc/src/main/groovy/org/apache/beam/gradle/BeamModulePlugin.groovy [hunk 1]
```diff
@@ -351,15 +351,6 @@ class BeamModulePlugin implements Plugin<Project> {
     Integer numParallelTests = 1
     // Whether the pipeline needs --sdk_location option
     boolean needsSdkLocation = false
-    // Categories for Java tests to run.
-    Closure javaTestCategories = {
-      includeCategories 'org.apache.beam.sdk.testing.UsesCrossLanguageTransforms'
-      // Use the following to include / exclude categories:
```
Comment (chamikaramj) [inline buildSrc/src/main/groovy/org/apache/beam/gradle/BeamModulePlugin.groovy:357]: Should this comment be removed ?

File: buildSrc/src/main/groovy/org/apache/beam/gradle/BeamModulePlugin.groovy [hunk 2]
```diff
@@ -2283,6 +2275,7 @@ class BeamModulePlugin implements Plugin<Project> {
         "java_expansion_service_jar": expansionJar,
         "java_port": javaPort,
         "java_expansion_service_allowlist_file": javaClassLookupAllowlistFile,
+        "python_expansion_service_fqn_glob": "\\*",
```
Comment (chamikaramj) [inline buildSrc/src/main/groovy/org/apache/beam/gradle/BeamModulePlugin.groovy:2278]: "python_expansion_service_allowlist" ?

File: examples/java/src/main/java/org/apache/beam/examples/multilang/PythonDataframeWordCount.java [hunk 3]
```diff
@@ -0,0 +1,152 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
```
Comment (chamikaramj) [inline examples/java/src/main/java/org/apache/beam/examples/multilang/PythonDataframeWordCount.java:2]: Can we add this to "examples/multi-language" instead ?
https://github.com/apache/beam/tree/master/examples/multi-language

File: examples/java/src/main/java/org/apache/beam/examples/multilang/PythonDataframeWordCount.java [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+ * from the input text file. The Python expansion service provided by --expansionService must allow
+ * the expansion of apache_beam.dataframe.transforms.DataframeTransform (which can be done by
+ * passing --fully_qualified_name_glob commandline option when launching the expansion service).
+ *
+ * <p>Note that, for using Dataflow Runner, you should specify the following two additional
+ * arguments:
+ *
+ * <pre>{@code
+ * --experiments=use_runner_v2
+ * --sdkHarnessContainerImageOverrides=.*python.*,gcr.io/apache-beam-testing/beam-sdk/beam_python3.8_sdk:latest
```
Comment (chamikaramj) [inline examples/java/src/main/java/org/apache/beam/examples/multilang/PythonDataframeWordCount.java:54]: In the following examples, I simplified by providing instructions for running with the latest released Beam version. Can we do the same here ? Users should be able to just run this example without additional setup (no need to startup an expansion service or push containers for example).

https://github.com/apache/beam/blob/master/sdks/python/apache_beam/examples/kafkataxi/README.md
https://github.com/apache/beam/tree/master/examples/multi-language

File: examples/java/src/main/java/org/apache/beam/examples/multilang/PythonDataframeWordCount.java [hunk 5]
```diff
@@ -0,0 +1,152 @@
...
+    /** Set this required option to specify where to write the output. */
+    @Description("Path of the file to write to")
+    @Required
+    String getOutput();
+
+    void setOutput(String value);
+
+    /** Set this required option to specify Python expansion service URL. */
+    @Description("URL of Python expansion service")
+    @Required
```
Comment (chamikaramj) [inline examples/java/src/main/java/org/apache/beam/examples/multilang/PythonDataframeWordCount.java:122]: Please make this option so that the example can just worth with the default expansion service.

File: runners/core-construction-java/src/test/java/org/apache/beam/runners/core/construction/ValidateRunnerXlangTest.java [hunk 6]
```diff
@@ -287,7 +287,11 @@ protected void pythonDependenciesTest(Pipeline pipeline) {
   @RunWith(JUnit4.class)
   public static class SingleInputOutputTest extends ValidateRunnerXlangTestBase {
     @Test
-    @Category({ValidatesRunner.class, UsesCrossLanguageTransforms.class})
+    @Category({
+      ValidatesRunner.class,
```
Comment (chamikaramj) [inline runners/core-construction-java/src/test/java/org/apache/beam/runners/core/construction/ValidateRunnerXlangTest.java:291]: Is there value in running both Java and Python pipelines with Java and Python expansion services ? I wonder if we can simplify (and hopefully reduce flakiness) by running Java tests only with Python expansion service and vice versa.

File: sdks/java/extensions/python/src/test/java/org/apache/beam/sdk/extensions/python/DataframeTransformTest.java [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+    Row foo1 = Row.withSchema(schema).withFieldValue("a", 100L).withFieldValue("b", 1).build();
+    Row foo2 = Row.withSchema(schema).withFieldValue("a", 100L).withFieldValue("b", 2).build();
+    Row foo3 = Row.withSchema(schema).withFieldValue("a", 100L).withFieldValue("b", 3).build();
+    Row bar4 = Row.withSchema(schema).withFieldValue("a", 200L).withFieldValue("b", 4).build();
+    PCollection<Row> col =
+        testPipeline
+            .apply(Create.of(foo1, foo2, bar4))
+            .setRowSchema(schema)
+            .apply(
+                PythonExternalTransform.<PCollection<Row>, PCollection<Row>>from(
```
Comment (chamikaramj) [inline sdks/java/extensions/python/src/test/java/org/apache/beam/sdk/extensions/python/DataframeTransformTest.java:99]: I think we should add a Java wrapper to simplify this for users (even though that might end up being a small class).

File: sdks/java/extensions/python/src/test/java/org/apache/beam/sdk/extensions/python/DataframeTransformTest.java [hunk 8]
```diff
@@ -0,0 +1,105 @@
...
+import org.junit.runners.JUnit4;
+
+@RunWith(JUnit4.class)
+public class DataframeTransformTest {
+  @Rule public transient TestPipeline testPipeline = TestPipeline.create();
+  private PipelineResult pipelineResult;
+  private static String expansionAddr;
+
+  @BeforeClass
+  public static void setUpClass() {
```
Comment (chamikaramj) [inline sdks/java/extensions/python/src/test/java/org/apache/beam/sdk/extensions/python/DataframeTransformTest.java:52]: Can these setup/teardown methods be moved to a common class instead of repeating in all Java x-lang tests ?

**Review by chamikaramj — CHANGES_REQUESTED** [review]
Thanks!

File: buildSrc/src/main/groovy/org/apache/beam/gradle/BeamModulePlugin.groovy (hunk 1 above)
Comment (ihji) [reply buildSrc/src/main/groovy/org/apache/beam/gradle/BeamModulePlugin.groovy:357]: `javaTestCategories` is removed from `CrossLanguageValidatesRunnerConfiguration`.  Consumers don't need to override the test categories.

File: buildSrc/src/main/groovy/org/apache/beam/gradle/BeamModulePlugin.groovy (hunk 2 above)
Comment (ihji) [reply buildSrc/src/main/groovy/org/apache/beam/gradle/BeamModulePlugin.groovy]: Changed it to `python_expansion_service_allowlist_glob`

File: examples/java/src/main/java/org/apache/beam/examples/multilang/PythonDataframeWordCount.java (hunk 3 above)
Comment (ihji) [reply examples/java/src/main/java/org/apache/beam/examples/multilang/PythonDataframeWordCount.java]: Moved.

File: examples/java/src/main/java/org/apache/beam/examples/multilang/PythonDataframeWordCount.java (hunk 5 above)
Comment (ihji) [reply examples/java/src/main/java/org/apache/beam/examples/multilang/PythonDataframeWordCount.java]: It's now optional but users need to specify it until we release a new SDK.

File: sdks/java/extensions/python/src/test/java/org/apache/beam/sdk/extensions/python/DataframeTransformTest.java (hunk 8 above)
Comment (ihji) [reply sdks/java/extensions/python/src/test/java/org/apache/beam/sdk/extensions/python/DataframeTransformTest.java]: Created a common base class.

File: runners/core-construction-java/src/test/java/org/apache/beam/runners/core/construction/ValidateRunnerXlangTest.java (hunk 6 above)
Comment (ihji) [reply runners/core-construction-java/src/test/java/org/apache/beam/runners/core/construction/ValidateRunnerXlangTest.java:237]: Tests using the same type of expansion service with the SDK will be skipped:
https://github.com/apache/beam/blob/bbc1867e6f3bcdb426e7f95ba923428017c0f151/buildSrc/src/main/groovy/org/apache/beam/gradle/BeamModulePlugin.groovy#L2356
https://github.com/apache/beam/blob/bbc1867e6f3bcdb426e7f95ba923428017c0f151/buildSrc/src/main/groovy/org/apache/beam/gradle/BeamModulePlugin.groovy#L2386

**Comment by chamikaramj** [comment]
Please let me know when this is ready for another look.

File: sdks/java/extensions/python/src/test/java/org/apache/beam/sdk/extensions/python/DataframeTransformTest.java (hunk 7 above)
Comment (ihji) [reply sdks/java/extensions/python/src/test/java/org/apache/beam/sdk/extensions/python/DataframeTransformTest.java]: Done.

**Comment by ihji** [comment]
Run XVR_Direct PostCommit

**Comment by ihji** [comment]
Run XVR_JavaUsingPython_Dataflow PostCommit

**Comment by ihji** [comment]
Run Java PreCommit

File: examples/java/src/main/java/org/apache/beam/examples/multilang/PythonDataframeWordCount.java (hunk 4 above)
Comment (ihji) [reply examples/java/src/main/java/org/apache/beam/examples/multilang/PythonDataframeWordCount.java]: Changed to use auto expansion service from PythonExternalTransform (However, the example will not work until we release a newer SDK version since PythonExternalTransform launches an expansion service by installing a released SDK)

File: examples/multi-language/src/main/java/org/apache/beam/examples/multilanguage/PythonDataframeWordCount.java [hunk 9]
```diff
@@ -42,19 +41,20 @@
...
- * arguments:
+ * <p>The example command below shows how to run this pipeline on Dataflow runner:
  *
  * <pre>{@code
- * --experiments=use_runner_v2
- * --sdkHarnessContainerImageOverrides=.*python.*,gcr.io/apache-beam-testing/beam-sdk/beam_python3.8_sdk:latest
+ * ./gradlew :examples:multi-language:pythonDataframeWordCount --args=" \
+ * --runner=DataflowRunner \
+ * --output=gs://{$OUTPUT_BUCKET}/count \
+ * --experiments=use_runner_v2 \
```
Comment (chamikaramj) [inline examples/multi-language/src/main/java/org/apache/beam/examples/multilanguage/PythonDataframeWordCount.java:52]: Is this required ? For Python if we detect multi-language transforms for DataflowRunner we add this experiment automatically (can be updated in a separate PR).

**Comment by chamikaramj** [comment]
Run XVR_Dataflow PostCommit

**Comment by chamikaramj** [comment]
Run XVR_PythonUsingJava_Dataflow PostCommit

**Review by chamikaramj — COMMENTED** [review]
LGTM. Thanks.

File: examples/multi-language/src/main/java/org/apache/beam/examples/multilanguage/PythonDataframeWordCount.java (hunk 9 above)
Comment (ihji) [reply examples/multi-language/src/main/java/org/apache/beam/examples/multilanguage/PythonDataframeWordCount.java:52]: I couldn't submit the job without `use_runner_v2` flag. We can look into it in a separate PR.

**Comment by ihji** [comment]
Run SQL PreCommit

**Comment by ihji** [comment]
Run SQL_Java17 PreCommit

**Comment by robertwb** [comment]
I'm curious why `sdks/java/extensions/python/src/test/java/org/apache/beam/sdk/extensions/python/io/DataframeTransform.java` is in `src/test/.../io` rather than `src/main/.../python`.

**Comment by ihji** [comment]
> I'm curious why `sdks/java/extensions/python/src/test/java/org/apache/beam/sdk/extensions/python/io/DataframeTransform.java` is in `src/test/.../io` rather than `src/main/.../python`.

@robertwb Ah, that's my mistake. Thanks for pointing out. I think it's fixed in Cham's Python Map PR.

**Comment by chamikaramj** [comment]
Yeah, I'm moving it to "sdks/java/extensions/python/src/main/java/org/apache/beam/sdk/extensions/python/transforms/" in https://github.com/apache/beam/pull/17696
